File: app/src/main/java/uk/co/sundroid/activity/location/SavedLocationsActivity.java
package uk.co.sundroid.activity.location;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.LinearLayout;
import android.widget.TextView;
import uk.co.sundroid.R;
import uk.co.sundroid.R.attr;
import uk.co.sundroid.R.id;
import uk.co.sundroid.R.layout;
import uk.co.sundroid.dao.DatabaseHelper;
import uk.co.sundroid.domain.LocationDetails;
import uk.co.sundroid.util.SharedPrefsHelper;
import uk.co.sundroid.util.geometry.Accuracy;

import java.util.List;

public class SavedLocationsActivity extends AbstractLocationActivity implements OnClickListener, DialogInterface.OnClickListener {

	private DatabaseHelper db;

    private static final int DIALOG_DELETE = 204;
    private static final int TAG_ACTION = attr.sundroid_custom_1;
    private static final int TAG_LOCATION_ID = attr.sundroid_custom_2;
	private static final String ACTION_DELETE = "delete";
    private static final String ACTION_VIEW = "view";

    private long contextSavedLocationId = 0;

	@Override
	protected int getLayout() {
		return R.layout.loc_saved;
	}

	@Override
	protected String getViewTitle() {
		return "Saved locations";
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		db = new DatabaseHelper(this);
        populateSavedLocations();
	}

	@Override
	protected void onDestroy() {
		super.onDestroy();
		if (db != null) {
			db.close();
		}
	}

    @Override
    public void onClick(View view) {
        String savedLocationId = (String)view.getTag(TAG_LOCATION_ID);
        String action = (String)view.getTag(TAG_ACTION);
        if (savedLocationId != null && action != null && db != null) {
            LocationDetails locationDetails = db.getSavedLocation(Integer.parseInt(savedLocationId));
            if (action.equals(ACTION_VIEW)) {
                SharedPrefsHelper.saveSelectedLocation(this, locationDetails);
                if (locationDetails.getTimeZone() == null) {
                    Intent intent = new Intent(getApplicationContext(), TimeZonePickerActivity.class);
                    intent.putExtra(TimeZonePickerActivity.INTENT_MODE, TimeZonePickerActivity.MODE_SELECT);
                    startActivityForResult(intent, TimeZonePickerActivity.REQUEST_TIMEZONE);
                } else {
                    setResult(LocationSelectActivity.RESULT_LOCATION_SELECTED);
                    finish();
                }
            } else if (action.equals(ACTION_DELETE)) {
                confirmDelete(Integer.parseInt(savedLocationId));
            }
            return;
        }
        super.onClick(view);
    }

    @Override
    public void onClick(DialogInterface dialogInterface, int button) {
        if (button == DialogInterface.BUTTON_POSITIVE) {
            delete();
        }
        dialogInterface.dismiss();
    }

    @Override
    public Dialog onCreateDialog(int id) {
        if (id == DIALOG_DELETE) {
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle("Delete");
            builder.setMessage("Delete this saved location?");
            builder.setPositiveButton("OK", this);
            builder.setNegativeButton("Cancel", this);
            return builder.create();
        }
        return super.onCreateDialog(id);
    }

    private void populateSavedLocations() {

        List<LocationDetails> locations = db.getSavedLocations();
        LinearLayout list = findViewById(id.savedLocationsList);
        list.removeAllViews();

        if (locations.size() > 0) {
            show(id.savedLocationsList);
            remove(id.savedLocationsNone);
        } else {
            remove(id.savedLocationsList);
            show(id.savedLocationsNone);
        }

        boolean first = true;
        for (final LocationDetails location : locations) {
            if (!first) {
                getLayoutInflater().inflate(layout.divider, list);
            }
            View row = View.inflate(this, layout.loc_saved_row, null);
            TextView name = row.findViewById(R.id.savedLocName);
            name.setText(location.getName());
            TextView coords = row.findViewById(R.id.savedLocCoords);
            coords.setText(location.getLocation().getPunctuatedValue(Accuracy.MINUTES));

            row.findViewById(id.savedLocText).setTag(TAG_LOCATION_ID, Integer.toString(location.getId()));
            row.findViewById(id.savedLocText).setTag(TAG_ACTION, ACTION_VIEW);
            row.findViewById(id.savedLocText).setOnClickListener(this);
            row.findViewById(id.savedLocDelete).setTag(TAG_LOCATION_ID, Integer.toString(location.getId()));
            row.findViewById(id.savedLocDelete).setTag(TAG_ACTION, ACTION_DELETE);
            row.findViewById(id.savedLocDelete).setOnClickListener(this);
            list.addView(row);

            first = false;
        }

    }

    private void confirmDelete(int savedLocationId) {
        contextSavedLocationId = savedLocationId;
        showDialog(DIALOG_DELETE);
    }

    private void delete() {
        if (contextSavedLocationId > 0 && db != null) {
            db.deleteSavedLocation(contextSavedLocationId);
            populateSavedLocations();
        }
    }

}

File: app/src/main/kotlin/uk/co/sundroid/activity/location/SavedLocationsActivity.kt
package uk.co.sundroid.activity.location

import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import uk.co.sundroid.R
import uk.co.sundroid.R.id
import uk.co.sundroid.domain.LocationDetails
import uk.co.sundroid.util.dao.DatabaseHelper
import uk.co.sundroid.util.prefs.Prefs
import uk.co.sundroid.util.geometry.Accuracy

class SavedLocationsActivity : AbstractLocationActivity() {

    private var db: DatabaseHelper? = null

    override val layout: Int
        get() = R.layout.loc_saved

    override val viewTitle: String
        get() = "Saved locations"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        db = DatabaseHelper(this)
        populateSavedLocations()
    }

    override fun onDestroy() {
        super.onDestroy()
        db?.close()
        db = null
    }

    private fun populateSavedLocations() {
        val db = this.db ?: return

        val locations = db.savedLocations
        val list = findViewById<LinearLayout>(id.savedLocationsList)
        list.removeAllViews()

        if (locations.size > 0) {
            show(id.savedLocationsList)
            remove(id.savedLocationsNone)
        } else {
            remove(id.savedLocationsList)
            show(id.savedLocationsNone)
        }

        var first = true
        for (location in locations) {
            if (!first) {
                layoutInflater.inflate(R.layout.divider, list)
            }
            val row = View.inflate(this, R.layout.loc_saved_row, null)
            row.findViewById<TextView>(R.id.savedLocName).text = location.name
            row.findViewById<TextView>(R.id.savedLocCoords).text = location.location.getPunctuatedValue(Accuracy.MINUTES)
            row.findViewById<View>(id.savedLocText).setOnClickListener { select(location) }
            row.findViewById<View>(id.savedLocDelete).setOnClickListener { confirmDelete(location) }
            list.addView(row)
            first = false
        }

    }

    private fun select(location: LocationDetails) {
        Prefs.saveSelectedLocation(this, location)
        if (location.timeZone == null) {
            val intent = Intent(this, TimeZonePickerActivity::class.java)
            intent.putExtra(TimeZonePickerActivity.INTENT_MODE, TimeZonePickerActivity.MODE_SELECT)
            startActivityForResult(intent, TimeZonePickerActivity.REQUEST_TIMEZONE)
        } else {
            setResult(LocationSelectActivity.RESULT_LOCATION_SELECTED)
            finish()
        }
    }

    private fun confirmDelete(location: LocationDetails) {
        val builder = AlertDialog.Builder(this)
        builder.setTitle("Delete")
        builder.setMessage("Delete this saved location?")
        builder.setPositiveButton("OK", { _, _ -> delete(location) })
        builder.setNegativeButton("Cancel", null)
        builder.create().show()
    }

    private fun delete(location: LocationDetails) {
        db?.deleteSavedLocation(location.id)
        populateSavedLocations()
    }

}
